sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,171 @@
+	watcher *inotify.Watcher
+	channel chan appMonitorData
+}
+
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appMonitorData),
+	watched: make(map[string][]*appMonitorData),
+}
+
+func onFilesDeleted(filename string) {
+	basePath := path.Dir(filename)
+	appWatchers := currentCGroupMonitor.watched[basePath]
+	var newList []*appMonitorData
+	for _, app := range appWatchers {
+		for _, folder := range app.cgroupPaths {
+			if folder == filename {
+				app.npaths--
+			}
+		}
+		if app.npaths == 0 {
+			// all the folders have disappeared, so notify that this app has no more instances running
+			app.channel <- app.name
+		} else {
+			newList = append(newList, app)
+		}
+	}
+	if len(newList) != 0 {
+		currentCGroupMonitor.watched[basePath] = newList
+	} else {
+		delete(currentCGroupMonitor.watched, basePath)
+		currentCGroupMonitor.watcher.RemoveWatch(basePath)
+	}
+}
+
+func onFilesAdded(newApp *appMonitorData) {
+	if newApp.npaths == 0 {
+		newApp.channel <- newApp.name
+		return
+	}
+	addedPaths := false
+	for _, fullPath := range newApp.cgroupPaths {
+		basePath := path.Dir(fullPath) // Monitor the path containing this folder
+		_, exists := currentCGroupMonitor.watched[basePath]
+		if exists {
+			continue
+		}
+		err := currentCGroupMonitor.watcher.AddWatch(basePath, inotify.InDelete)
+		if err != nil {
+			continue
+		}
+		if _, err := os.Stat(fullPath); errors.Is(err, os.ErrNotExist) {
+			// if the file/folder to monitor doesn't exist after the parent being added, remove it
+			currentCGroupMonitor.watcher.RemoveWatch(basePath)
+			continue
+		}
+		currentCGroupMonitor.watched[basePath] = append(currentCGroupMonitor.watched[basePath], newApp)
+		addedPaths = true
+	}
+	if !addedPaths {
+		// if the files/folders to monitor don't exist, send the notification now
+		newApp.channel <- newApp.name
+	}
+}
+
+func monitorMainLoop() {
+	for {
+		select {
+		case event := <-currentCGroupMonitor.watcher.Event:
+			if event.Mask&inotify.InDelete != 0 {
+				onFilesDeleted(event.Name)
+			}
+		case newApp := <-currentCGroupMonitor.channel:
+			onFilesAdded(&newApp)
+		}
+	}
+}
+
+// DefaultCGroupMonitor launches the main loop and returns the CGroup singleton
+func DefaultCGroupMonitor() *CGroupMonitor {
+	if currentCGroupMonitor.watcher == nil {
+		wd, err := inotify.NewWatcher()
+		if err != nil {
+			return nil
+		}
+		currentCGroupMonitor.watcher = wd
+		go monitorMainLoop()
+	}
+	return &currentCGroupMonitor
+}
+
+// MonitorSnap is the method to call to monitor the running instances of an specific Snap.
+// It receives the name of the snap to monitor (for example, "firefox" or "steam")
+// and a channel. The caller can wait on the channel, and when all the instances of
+// the specific snap have ended, the name of the snap will be sent through the channel.
+// This allows to use the same channel to monitor several snaps
+func (this CGroupMonitor) MonitorSnap(snapName string, channel chan string) {
```

Comment:
We usually don't use `this` as variable name (though it's fine for me if you keep it as it is), but you definitely need to change all the methods to be 
```suggestion
func (this *CGroupMonitor) MonitorSnap(snapName string, channel chan string) {
```
You can check the assembly at https://godbolt.org/z/n1914d8dr with and without the `*` to see what difference it makes. :-)

Reply:
I see... If I understand it right, without the * it passes a copy of the structure/object, while with the * it passes a reference.

Done.